File: backend/src/routes/incydenty.js
'use strict';
const express = require('express');
const { body } = require('express-validator');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { validate } = require('../middleware/validate');
const { writeAudit } = require('../services/auditService');
const { notifyIT } = require('../services/notificationService');
const { sendXlsxResponse, sendCsvResponse } = require('../services/excelService');

const router = express.Router();

router.get('/', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, status, format } = req.query;
  let q = `SELECT ri.*, t.nazwa as tenant_nazwa,
                  u.imie || ' ' || u.nazwisko as zglaszajacy_nazwa
             FROM rejestr_incydentow ri
             JOIN tenants t ON t.id = ri.tenant_id
             JOIN uzytkownicy u ON u.id = ri.zglaszajacy_id
            WHERE 1=1`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND ri.tenant_id=$${params.length}`; }
  if (status) { params.push(status); q += ` AND ri.status=$${params.length}`; }
  q += ' ORDER BY ri.data_zgloszenia DESC';
  const { rows } = await db.query(q, params);

  const cols = [
    { header: 'Tytuł', key: 'tytul', width: 30 },
    { header: 'Jednostka', key: 'tenant_nazwa', width: 20 },
    { header: 'Typ', key: 'typ', width: 22 },
    { header: 'Poziom', key: 'poziom', width: 10 },
    { header: 'Status', key: 'status', width: 12 },
    { header: 'Data zgłoszenia', key: 'data_zgloszenia', width: 18 },
    { header: 'NIS2', key: 'dotyczy_nis2', width: 8 },
    { header: 'Zgłaszający', key: 'zglaszajacy_nazwa', width: 22 },
  ];
  if (format === 'xlsx') return sendXlsxResponse(res, cols, rows, 'Rejestr_incydentow', 'Incydenty');
  if (format === 'csv') return sendCsvResponse(res, cols, rows, 'Rejestr_incydentow');
  res.json(rows);
});

router.get('/:id', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { rows } = await db.query(
    `SELECT ri.*, t.nazwa as tenant_nazwa, u.imie || ' ' || u.nazwisko as zglaszajacy_nazwa
       FROM rejestr_incydentow ri
       JOIN tenants t ON t.id = ri.tenant_id
       JOIN uzytkownicy u ON u.id = ri.zglaszajacy_id
      WHERE ri.id=$1`,
    [req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Incydent nie istnieje.' });
  res.json(rows[0]);
});

router.post('/', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'),
  [
    body('tenant_id').isInt(),
    body('tytul').notEmpty().trim(),
    body('typ').isIn(['NIEAUTORYZOWANY_DOSTEP', 'NADUZYCIE_UPRAWNIEN', 'INNE']),
    body('poziom').isIn(['NISKI', 'SREDNI', 'WYSOKI', 'KRYTYCZNY']),
  ],
  validate,
  async (req, res) => {
    const { tenant_id, tytul, opis, typ, poziom, dotyczy_nis2 } = req.body;
    const { rows } = await db.query(
      `INSERT INTO rejestr_incydentow (tenant_id, tytul, opis, typ, poziom, zglaszajacy_id, dotyczy_nis2)
       VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING *`,
      [tenant_id, tytul, opis || null, typ, poziom, req.user.userId, dotyczy_nis2 || false]
    );
    await notifyIT({ typ: 'NOWY_INCYDENT', tresc: `Nowy incydent: ${tytul} (${poziom})`, link: `/incydenty/${rows[0].id}` });
    await writeAudit({ tenantId: tenant_id, userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'INCYDENT_UTWORZONO', tabelaDocelowa: 'rejestr_incydentow', rekordId: rows[0].id, noweDane: rows[0] });
    res.status(201).json(rows[0]);
  }
);

router.put('/:id', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tytul, opis, typ, poziom, status, dzialania_naprawcze, dotyczy_nis2 } = req.body;
  const dataZamkniecia = status === 'ZAMKNIETY' ? 'NOW()' : 'NULL';
  const { rows } = await db.query(
    `UPDATE rejestr_incydentow SET tytul=$1, opis=$2, typ=$3, poziom=$4, status=$5,
       dzialania_naprawcze=$6, dotyczy_nis2=$7,
       data_zamkniecia = CASE WHEN $5='ZAMKNIETY' THEN NOW() ELSE data_zamkniecia END
     WHERE id=$8 RETURNING *`,
    [tytul, opis || null, typ, poziom, status, dzialania_naprawcze || null, dotyczy_nis2 || false, req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Incydent nie istnieje.' });
  res.json(rows[0]);
});

router.delete('/:id', authenticate, roleGuard('SUPERADMIN'), async (req, res) => {
  await db.query('DELETE FROM rejestr_incydentow WHERE id=$1', [req.params.id]);
  res.json({ message: 'Incydent usunięty.' });
});

module.exports = router;

File: backend-py/routes/incydenty.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, require_roles, write_audit, get_client_ip
from services.notification_service import notify_it

router = APIRouter()
IT = ("IT_ADMIN", "SUPERADMIN")


class IncydentIn(BaseModel):
    tenant_id: int
    tytul: str
    opis: Optional[str] = None
    typ: str
    poziom: str = "NISKI"
    dotyczy_nis2: Optional[bool] = False


class IncydentUpdate(BaseModel):
    tytul: str
    opis: Optional[str] = None
    typ: str
    poziom: str
    status: str
    dzialania_naprawcze: Optional[str] = None
    dotyczy_nis2: Optional[bool] = False


@router.get("/")
async def list_incydenty(tenant_id: Optional[int] = None, status: Optional[str] = None,
                          format: Optional[str] = None,
                          session: Session = Depends(get_session),
                          current_user: CurrentUser = Depends(require_roles(*IT))):
    q = """SELECT ri.*, t.nazwa as tenant_nazwa, u.imie || ' ' || u.nazwisko as zglaszajacy_nazwa
             FROM rejestr_incydentow ri
             JOIN tenants t ON t.id=ri.tenant_id
             JOIN uzytkownicy u ON u.id=ri.zglaszajacy_id
            WHERE 1=1"""
    params: dict = {}
    if tenant_id:
        q += " AND ri.tenant_id=:tid"
        params["tid"] = tenant_id
    if status:
        q += " AND ri.status=:s"
        params["s"] = status
    q += " ORDER BY ri.data_zgloszenia DESC"
    rows = session.execute(text(q), params).mappings().all()
    data = [dict(r) for r in rows]

    if format in ("xlsx", "csv"):
        from services.excel_service import send_xlsx, send_csv
        cols = [
            {"header": "Tytuł", "key": "tytul"},
            {"header": "Jednostka", "key": "tenant_nazwa"},
            {"header": "Typ", "key": "typ"},
            {"header": "Poziom", "key": "poziom"},
            {"header": "Status", "key": "status"},
            {"header": "Data zgłoszenia", "key": "data_zgloszenia"},
            {"header": "NIS2", "key": "dotyczy_nis2"},
            {"header": "Zgłaszający", "key": "zglaszajacy_nazwa"},
        ]
        if format == "xlsx":
            return send_xlsx(data, cols, "Rejestr_incydentow", "Incydenty")
        return send_csv(data, cols, "Rejestr_incydentow")

    return data


@router.get("/{incydent_id}")
async def get_incydent(incydent_id: int, session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("SELECT ri.*, t.nazwa as tenant_nazwa, u.imie || ' ' || u.nazwisko as zglaszajacy_nazwa FROM rejestr_incydentow ri JOIN tenants t ON t.id=ri.tenant_id JOIN uzytkownicy u ON u.id=ri.zglaszajacy_id WHERE ri.id=:id"),
        {"id": incydent_id},
    ).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Incydent nie istnieje.")
    return dict(row)


@router.post("/", status_code=201)
async def create_incydent(body: IncydentIn, request: Request,
                           session: Session = Depends(get_session),
                           current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("INSERT INTO rejestr_incydentow (tenant_id,tytul,opis,typ,poziom,zglaszajacy_id,dotyczy_nis2,data_zgloszenia) VALUES (:tid,:t,:o,:typ,:p,:uid,:n,:dt) RETURNING *"),
        {"tid": body.tenant_id, "t": body.tytul, "o": body.opis, "typ": body.typ,
         "p": body.poziom, "uid": current_user.userId, "n": body.dotyczy_nis2 or False,
         "dt": datetime.utcnow()},
    ).mappings().first()
    session.commit()
    result = dict(row)
    notify_it("NOWY_INCYDENT", f"Nowy incydent: {body.tytul} ({body.poziom})",
              f"/incydenty/{result['id']}", session)
    write_audit(session, tenant_id=body.tenant_id, user_id=current_user.userId,
                username=current_user.username, rola=current_user.rola,
                akcja="INCYDENT_UTWORZONO", tabela_docelowa="rejestr_incydentow", rekord_id=result["id"],
                nowe_dane=result, ip_adres=get_client_ip(request))
    return result


@router.put("/{incydent_id}")
async def update_incydent(incydent_id: int, body: IncydentUpdate,
                           session: Session = Depends(get_session),
                           current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("""UPDATE rejestr_incydentow SET tytul=:t, opis=:o, typ=:typ, poziom=:p, status=:s,
                dzialania_naprawcze=:d, dotyczy_nis2=:n,
                data_zamkniecia=CASE WHEN :s='ZAMKNIETY' THEN :dt ELSE data_zamkniecia END
                WHERE id=:id RETURNING *"""),
        {"t": body.tytul, "o": body.opis, "typ": body.typ, "p": body.poziom, "s": body.status,
         "d": body.dzialania_naprawcze, "n": body.dotyczy_nis2 or False,
         "dt": datetime.utcnow(), "id": incydent_id},
    ).mappings().first()
    session.commit()
    if not row:
        raise HTTPException(status_code=404, detail="Incydent nie istnieje.")
    return dict(row)


@router.delete("/{incydent_id}")
async def delete_incydent(incydent_id: int, session: Session = Depends(get_session),
                           current_user: CurrentUser = Depends(require_roles("SUPERADMIN"))):
    session.execute(text("DELETE FROM rejestr_incydentow WHERE id=:id"), {"id": incydent_id})
    session.commit()
    return {"message": "Incydent usunięty."}
